### ISSUE TITLE
Wrong CPU clock detected by the calibration loop on my SKL

### ISSUE THREAD

**Issue by pcordes** [travisdowns/bimodal-performance#1]
[As discussed in SO comments](https://stackoverflow.com/questions/47851120/unexpectedly-poor-and-bimodal-performance-for-simple-store-loop-on-intel-skylake?noredirect=1#comment82677285_47851120), it detects 1.3GHz on my CPU.  That's exactly 1/3rd of the actual clock speed, 3.9GHz.

So for some reason the loop is running slower than expected, even though gcc7.1.1 -O1 did compile it to the expected 

```
+132>   mov    QWORD PTR [rip+0x200a8e],rax        # 0x5555557560f0
+139>   sub    rax,0x1
+143>   jne    0x55555555565b <CycleTimer::impl::Calibration<std::chrono::_V2::steady_clock, 10000ul, 11ul, 100000ul>::getGHz()+132> 
```

After a warmup of 100000 outer-loop iters, I get

```
(gdb) p results
$4 = {
  _M_elems = {7685, 7685, 7685, 7686, 7686, 7687, 7687, 7688, 7689, 7689, 7694}
}
```

The CPU is at 3.9GHz the whole time.

**Comment by travisdowns** [comment]
@pcordes - this is a real mystery. I've never seen it fail like that, and on other machines it has always produced the "expected" results (including too-low readings if using `powersave` governor and not enough `WARMUP` - but that always behaved in a fairly regular way, i.e., very visibly ramping up to max frequency).

I'm going to add two more calibtraion methods and I'd love if you'd test them.

**Comment by pcordes** [comment]
An asm version of the same loop does of course run at full speed (~1 store per clock, 2.99 IPC) in my usual static-binary test harness (with many more iterations):

```
00000000004000e0 <_start.loop>:
  4000e0:       48 89 2d 59 00 20 00    mov    QWORD PTR [rip+0x200059],rbp        # 600140 <buf>
  4000e7:       ff cd                   dec    ebp
  4000e9:       7f f5                   jg     4000e0 <_start.loop>
```

And sure, let me know when to pull and test.

**Comment by travisdowns** [comment]
Are you sure you aren't living in a simulated world?

**Comment by travisdowns** [comment]
@pcordes - can you check out down the [calibration branch](https://github.com/travisdowns/bimodal-performance/tree/calibration) and build & run?

It has four calibration methods, which will be output when you run in any mode, like:

```
$ ./weirdo-main asm
Estimated CPU speed ( store-asm):  3.49 GHz
Estimated CPU speed (store-asm2):  3.49 GHz
Estimated CPU speed (   add-asm):  3.49 GHz
Estimated CPU speed (   store-c):  3.49 GHz
```

Make sure you sync to at least 047273d68a1404f8d4f208b7f6a48a1d505b6585 since the verion had a copy/paste error that made all the asm calibrations use the "add" loop.

**Comment by pcordes** [comment]
`ITERS = 10000000, size_t TRIES = 11, size_t WARMUP = 2`  measures 3.90GHz.  I guess either something slow is apparently happening when we first enter the real non-warmup loop, or it runs slowly for a while.

gcc doesn't align its loops, and the warmup loop had the sub / jcc separated by a 16-byte boundary. But not a 32-byte boundary, I think.  I tried a misalignment like that in my `testloop.asm` with no change at all.  I forgot to notice the 2nd hex digit, so IDK if it was a meaningful misalignment or not.

Anyway, I wonder if there might be uop-cache effects, for example if the loop spans two uop-cache lines.  (My SKL has the microcode update that disables the loop buffer.)

I didn't go much further with the testing different misalignments because the specific surrounding code is going to matter as well as just the alignment.  IDK if the uop cache adapts by re-decoding a loop so it fits in one line, or if it is possible to create a tiny loop that runs at one per 2 clocks because of a uop-cache bottleneck on Skylake with no LSD.

**Comment by travisdowns** [comment]
> I tried a misalignment like that in my testloop.asm.

What happened?

Weird, the warmup is written so that the warmup and non-warmup iterations are identical, i.e., they use the same loop in the binary. That is, there is no warmup loop at all: we just do the main outer loop WARMUP + 1 times, and keep overwriting the results array. The last iteration is where the results are taken from. So it's not the classic:

```
for (int i = 0; i < WARMUP; i++) {
  // warmup
}
// for real now
```

**Comment by travisdowns** [comment]
@pcordes - the new "asm" calibration methods are always 32-byte aligned at least in the inner calibration loop, so running those may give us some more info.

If you have a moment I'd be interested in a gist of the `objdump -d` for your compiled `getGHzImpl` method.

**Comment by pcordes** [comment]
Updated my previous comment: there was no effect from splitting a 16-byte but not 32-byte boundary, as expected (derp).

So why are there two loops in the asm?  I noticed your source, but is gcc peeling the last outer-loop iteration and making a second timing loop?

Currently wrestling with the Makefile's `LDFLAGS`, because my gcc has `-pie` enabled by default.  I thought you could `LDFLAGS += -no-pie`, but it's not using it.

```
/usr/bin/ld: weirdo.o: relocation R_X86_64_32 against `.data' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Nonrepresentable section on output
```

Anyway, I could just run gcc manually to link, or tweak your asm with a `DEFAULT REL` directive.

**Comment by pcordes** [comment]
Oh, you had an explicit `g++` command line without `$(LDFLAGS)`.  I set `CC=g++` and then my LDFLAGS worked.  (Equally dirty hack, I guess. :P  I forget how you're supposed to tell make to link C++ object files.)

Anyway, with `cycle-timer.hpp` unmodified (`ITERS = 10000, size_t TRIES = 11, size_t WARMUP = 1000`), it all works fine now.

```
Estimated CPU speed ( store-asm):  3.90 GHz
Estimated CPU speed (store-asm2):  3.90 GHz
Estimated CPU speed (   add-asm):  3.90 GHz
Estimated CPU speed (   store-c):  3.89 GHz
```

This is leaning more and more towards an alignment issue of some sort.  I'll go back to the source that had the bug, and add a `.p2align` in the compiler output.

**Comment by travisdowns** [comment]
@pcordes - pull request is already accepted (and I rebased the `calibration` branch, so you might want to re-pull). I'll change the asm to store to the stack which will solve that.

Which two loops are you referring to? A copy/paste can help. 

Here's a snippet from my compiled `main` (everything gets inlined):

```
  400ed1:       80 3d 20 22 20 00 00    cmp    BYTE PTR [rip+0x202220],0x0        # 6030f8 <guard variable for CycleTimer::impl::Calibration<std::chrono::_V2::steady_clock, 10000ul, 11ul, 1000ul>::
getGHz()::ghz>
  400ed8:       0f 85 74 03 00 00       jne    401252 <main+0x50e>
  400ede:       bf f8 30 60 00          mov    edi,0x6030f8
  400ee3:       e8 e8 fb ff ff          call   400ad0 <__cxa_guard_acquire@plt>
  400ee8:       85 c0                   test   eax,eax
  400eea:       0f 84 62 03 00 00       je     401252 <main+0x50e>
  400ef0:       48 c7 44 24 10 e9 03    mov    QWORD PTR [rsp+0x10],0x3e9
  400ef7:       00 00 
  400ef9:       4c 8d 6c 24 40          lea    r13,[rsp+0x40]
  400efe:       4d 8d 7d 58             lea    r15,[r13+0x58]
  400f02:       e8 e9 fc ff ff          call   400bf0 <std::chrono::_V2::steady_clock::now()@plt>
  400f07:       49 89 c6                mov    r14,rax
  400f0a:       bf 10 27 00 00          mov    edi,0x2710
  400f0f:       e8 8c 0c 00 00          call   401ba0 <store_calibration_asm>
  400f14:       e8 d7 fc ff ff          call   400bf0 <std::chrono::_V2::steady_clock::now()@plt>
  400f19:       48 89 c3                mov    rbx,rax
  400f1c:       bf 20 4e 00 00          mov    edi,0x4e20
  400f21:       e8 7a 0c 00 00          call   401ba0 <store_calibration_asm>
  400f26:       e8 c5 fc ff ff          call   400bf0 <std::chrono::_V2::steady_clock::now()@plt>
  400f2b:       48 29 d8                sub    rax,rbx
  400f2e:       4c 29 f3                sub    rbx,r14
  400f31:       48 29 d8                sub    rax,rbx
  400f34:       49 89 45 00             mov    QWORD PTR [r13+0x0],rax
  400f38:       49 83 c5 08             add    r13,0x8
  400f3c:       4d 39 ef                cmp    r15,r13
  400f3f:       75 c1                   jne    400f02 <main+0x1be>
  400f41:       48 83 6c 24 10 01       sub    QWORD PTR [rsp+0x10],0x1
  400f47:       75 b0                   jne    400ef9 <main+0x1b5>
  400f49:       48 83 ec 08             sub    rsp,0x8
  400f4d:       6a 00                   push   0x0
  400f4f:       ba 06 00 00 00          mov    edx,0x6
  400f54:       48 8b 5c 24 40          mov    rbx,QWORD PTR [rsp+0x40]
  400f59:       48 89 de                mov    rsi,rbx
  400f5c:       48 8d 7c 24 50          lea    rdi,[rsp+0x50]
  400f61:       e8 3b 06 00 00          call   4015a1 <void std::__introsort_loop<long*, long, __gnu_cxx::__ops::_Iter_less_iter>(long*, long*, long, __gnu_cxx::__ops::_Iter_less_iter)>
```

There is only one main loop there, the calibration method, in this case `store_calibration_asm` is called twice because I'm doing a "delta" between two calibration loops, the second with twice the number of iterations, to cancel out constant factors like the function call and timer overhead.

**Comment by travisdowns** [comment]
> Oh, you had an explicit g++ command line without $(LDFLAGS). I set CC=g++ and then my LDFLAGS worked. (Equally dirty hack, I guess. :P I forget how you're supposed to tell make to link C++ object files.)

I don't think `make` has a rule to link C++ binaries, it only lists a rule of the form `$(CC) $(LDFLAGS) n.o $(LOADLIBES) $(LDLIBS)` for C binaries (see "Linking a single object file" [here](https://www.gnu.org/software/make/manual/html_node/Catalogue-of-Rules.html#Catalogue-of-Rules)). The problem is that `.c` and `.cpp` both produce `.o` files and so then there is no way make can distinguish the C and C++ cases for linking (which take `.o` files), so you end up only with the C rule, and that's why I wrote my own rule.

You are right I forgot to reproduce the form of the original built-in rule though, will fix.

**Comment by pcordes** [comment]
You got the backticks wrong in your previous comment >.<

Two loops:  with the old version of the source (before your update to add 4 calibration methods).  I still had it open in GDB.  Notice that they use different initial counter values, but both store to `sink`.  I didn't really try to follow the asm in detail to see what's going on, but here it is:

```
   0x000055555555564a <+119>:   mov    eax,0x989680     # 10000000 (from my larger ITERS source change)
   0x000055555555564f <+124>:   mov    QWORD PTR [rip+0x200a9a],rax        # 0x5555557560f0 <_ZN10CycleTimer4impl11CalibrationINSt6chrono3_V212steady_clockELm10000000ELm11ELm2EE4sinkE>
   0x0000555555555656 <+131>:   sub    rax,0x1
   0x000055555555565a <+135>:   jne    0x55555555564f <CycleTimer::impl::Calibration<std::chrono::_V2::steady_clock, 10000000ul, 11ul, 2ul>::getGHz()+124>
   0x000055555555565c <+137>:   call   0x555555554c60 <_ZNSt6chrono3_V212steady_clock3nowEv@plt>
   0x0000555555555661 <+142>:   mov    r14,rax
   0x0000555555555664 <+145>:   mov    eax,0x1312d00    # 20000000
   0x0000555555555669 <+150>:   mov    QWORD PTR [rip+0x200a80],rax        # 0x5555557560f0 <_ZN10CycleTimer4impl11CalibrationINSt6chrono3_V212steady_clockELm10000000ELm11ELm2EE4sinkE>
=> 0x0000555555555670 <+157>:   sub    rax,0x1
   0x0000555555555674 <+161>:   jne    0x555555555669 <CycleTimer::impl::Calibration<std::chrono::_V2::steady_clock, 10000000ul, 11ul, 2ul>::getGHz()+150>
   0x0000555555555676 <+163>:   call   0x555555554c60 <_ZNSt6chrono3_V212steady_clock3nowEv@plt>
  ... use the time result
```

**Comment by pcordes** [comment]
Oh I see where the 2 loops come from in the source.  The source does `CAL(ITERS);` twice in the TRIES loop to account for measurement overhead.

But if some weird front-end effect slowed down the 2nd loop by a factor of 2, that would give us a factor of 3 difference in measured performance.  A factor of 2 in an actual loop is easy to explain as a uop-cache effect, and it's only the math that's magnifying this to a factor of 3 because of unfortunately incorrect assumptions that 2 copies of the same loop will always run at the same speed.

Using calc to check my math with a simple example of 100 time units, and variable speed-factors for the 2 loops:

```
define f(a, b) { local t0 = 0; local t1 = 100*a; local t2 = t1 + 2*100*b;  print (t2 - t1) - (t1 - t0); }
f(a,b) redefined
; f(1,1)
100 /* 0x64 */
; f(1,2)
300 /* 0x12c */
```

**Comment by travisdowns** [comment]
@pcordes - yes, as I mentioned above because I'm doing a "delta" measurement there are two source-level calls to the "cal" method, and in the case of the visible `store_calibration` method those are inlined resuling in two separate source level innermost loops (I had thought you were saying there were two loops for one of the outer `TRIES` or `WARMUP+1` loops). The asm ones don't that problem since we always just up with a `call`.

I'll add "noinline" and "align function" attributes to avoid this problem in the future.

**Comment by travisdowns** [comment]
> You got the backticks wrong in your previous comment >.<

Weird, you mean [in this comment](https://github.com/travisdowns/bimodal-performance/issues/1#issuecomment-352276092)? I corrected it immediately, you must have refreshed at a very specific time!

**Comment by travisdowns** [comment]
> incorrect assumptions that 2 copies of the same loop will always run at the same speed.

I should add a sanity check that the times are "reasonable", i.e., that the second loop is about twice as long as the first, at least within some very small tolerance. I use this same approach in [uarch-bench](https://github.com/travisdowns/uarch-bench) for the calibration (if you aren't using performance counter mode), so it's worth getting it right. When it works it is generally very accurate (i.e., a variance of less than 0.1%, often around 0.01% - much of it attributable to using wall-clock timers that have periodic adjustments to keep them in sync with network time).

You actually want to keep `ITERATIONS` to a smallish number, since then most of the runs will have zero interrupts and the median will pick out the "true" clock speed. This runs contrary to the usual idea to just run lots of iterations to "average out" the random deviations (this doesn't work when the deviations are biased in one direction, which is usually the case here).

**Comment by travisdowns** [comment]
@pcordes This should be fixed in baed9494c4bdd5c3f09734e01233f89ad05207b5 on master. I set the `CAL` method alignment to 32 and disabled inlining of that function.

If you have a chance, checkout master and see if the 1.3 has changed to 3.9.
